migrator/src/main/scala/com/scylladb/migrator/readers/PartitionMetadataReader.scala

```diff
@@ -0,0 +1,87 @@
+package com.scylladb.migrator.readers
+
+import org.apache.log4j.LogManager
+import org.apache.spark.sql.{ DataFrame, SparkSession }
+import org.apache.spark.sql.functions._
+
+case class PartitionMetadata(
+  partitionId: Int,
+  filename: String
+)
+
+/**
+  * This reader uses Spark's internal partition information to build mappings
+  * between partition IDs and file paths. This allows us to track when all
+  * partitions of a file have been processed, enabling file-level savepoints.
+  */
+object PartitionMetadataReader {
+  private val logger = LogManager.getLogger("com.scylladb.migrator.readers.PartitionMetadataReader")
+
+  def readMetadata(spark: SparkSession, filePaths: Seq[String]): Seq[PartitionMetadata] = {
+    logger.info(s"Reading partition metadata from ${filePaths.size} file(s)")
+    val df = spark.read.parquet(filePaths: _*)
+    readMetadataFromDataFrame(df)
+  }
+
+  def readMetadataFromDataFrame(df: DataFrame): Seq[PartitionMetadata] =
+    try {
+
+      val partitionInfo = df
+        .select(input_file_name().as("filename"))
+        .rdd
+        .mapPartitionsWithIndex { (partitionId, iter) =>
+          val files = iter.map(row => row.getString(0)).toSet
+          files.map(filename => (partitionId, filename)).iterator
```

Comment:
how does this behave when you have lots of 100-200MB files and few files that are around 5G (how is splitting)